### ISSUE TITLE
AttributeError: 'Request' object has no attribute 'log' on TestCase

### ISSUE THREAD

**Issue by bkovacev** [aschn/drf-tracking#53]
Hello, I have stumbled upon the same thing on my test suite. I'm testing accessing the given resource without authentication.

```python
class DriverReadFailureTests(APITestCase):
    def setUp(self):
        self.client = APIClient()
        self.user = User.objects.create_superuser('admin', 'admin@admin.com', 'admin123')
        self.client.force_login(user=self.user)
        self.token = Token.objects.create(user=self.user)
        self.client.post('/api/drivers/', data=driver_data, format='json')
        self.client.post('/api/drivers/', data=driver_data_second, format='json')
        self.client.logout()

    def test_read_no_auth_driver(self):
        """Try to retrieve json representation of previously added driver.
        """
        response = self.client.get('/api/drivers/')
        self.assertEqual(response.status_code, 401)
```
Traceback:
Error

```shell
Traceback (most recent call last):
  File "/home/bkovacev/sally-projects/ops4/drivers/tests.py", line 316, in test_read_no_auth_driver
    response = self.client.get('/api/drivers/')
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/test.py", line 282, in get
    response = super(APIClient, self).get(path, data=data, **extra)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/test.py", line 208, in get
    return self.generic('GET', path, **r)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/test/client.py", line 409, in generic
    return self.request(**r)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/test.py", line 279, in request
    return super(APIClient, self).request(**kwargs)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/test.py", line 231, in request
    request = super(APIRequestFactory, self).request(**kwargs)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/test/client.py", line 494, in request
    six.reraise(*exc_info)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/core/handlers/exception.py", line 42, in inner
    response = get_response(request)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/views.py", line 483, in dispatch
    response = self.handle_exception(exc)
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework_tracking/mixins.py", line 80, in handle_exception
    self.request.log.errors = traceback.format_exc()
  File "/home/bkovacev/envs/ops4-env/local/lib/python2.7/site-packages/rest_framework/request.py", line 384, in __getattribute__
    six.reraise(info[0], info[1], info[2].tb_next)
  File "<string>", line 2, in reraise
AttributeError: 'Request' object has no attribute 'log'

```
How do I approach this?

**Comment by avelis** [comment]
@bkovacev Thanks for writing in. It looks like this an interesting fallout of two PR's https://github.com/aschn/drf-tracking/commit/210511e4ac66be55d91c56c22fa3e5643bee320e & https://github.com/aschn/drf-tracking/commit/ab25cb73123751cf4821488ea12ab241cd3ba5f9.

One adds error tracking and the other is to gracefully handle APIRequestLog saving failures. In your scenario when the handle_exception get's called it assumes the log instance is attached to the request object and it clearly isn't. It isn't even there.

Quite the bug. I don't have an immediate solution other than determining that why the APIRequestLog isn't saving during the mixins initial method. I would need to write up a test and come up with a solution to possibly gracefully not save in handle_exception if one doesn't exist.

Also open to a PR if you have the time to address this. Will look into this sometime by next week.

**Comment by bkovacev** [comment]
Thank you for the reply! Let me see what I can do about it this weekend 👍 I'll keep you posted!

**Comment by vince-stark** [comment]
This bug is also important for another reason - when a logged in user session expires, it triggers the same error, instead of following DRF's authentication scheme, and returning 403 where needed.

So, @bkovacev @avelis any updates? I can help as needed!

**Comment by avelis** [comment]
@vince-stark I will say I am a little pressed for time to come up change to fix this. If you have the time. the issue I believe is the side effect of a failed initialized log (with a returning) and the handle_exception is called and no log attribute exists. This is the scenario that needs a unit test & fix.

**Comment by vince-stark** [comment]
@avelis sounds good - I'll work on it 👍

**Comment by avelis** [comment]
@vince-stark Thanks again for taking an interest in helping out the project. I am always accepting of contributions to make this library better.

**Comment by triat** [comment]
@avelis I think this one is fixed (if I check the PR which contains the commit #54 )

**Comment by avelis** [comment]
This issue should be addressed in v1.3.1
